### PR TITLE
Several APC bug fixes

### LINKED ISSUE #7961: Rebuilt APCs don't clear their respective power alarm
APCs smashed by nightmares or other miscreants don't clear their power alarm after being rebuilt. 
## Reproduction:
Smash an APC till it breaks, smash it further to open the cover, deconstruct it, reconstruct it and voila! The lights are back on but the power alarm is still there.

image

### LINKED ISSUE #8244: APCs hacked by an AI do not look emagged.
## Round ID & Round Date:

41926
23/12/2022
## Testmerges:
[#8231: 'Revert "Bumps Auxmos from 1.1.2 to 2.2.1 and BYOND to 514.1589 (#7959)"' by KubeRoot at commit 96f7ae89d4](https://www.github.com/beestation/beestation-hornet/pull/8231)
[#8201: 'Adds the ability to see most poll types' by PowerfulBacon at commit fb25c679e4](https://www.github.com/beestation/beestation-hornet/pull/8201)

## Reproduction:
Get a combat upgraded or malfunctioning AI to hack an APC. 
Look at the APC.
It does not look emagged.

### PR DESCRIPTION
**Pull request by CydiaLamiales**
## About The Pull Request
* closes https://github.com/BeeStation/BeeStation-Hornet/issues/8244 
* closes https://github.com/BeeStation/BeeStation-Hornet/issues/7961

Also makes smashed APCs that weren't fully destroyed report a power alarm in their area.

## Why It's Good For The Game
Less bugs good, APCs being easier to repair after a Nightmare went ham on them is also good.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

Malf AI hacking an APC and it becoming blue : 

https://files.catbox.moe/pzg1b8.mp4

APC being smashed, deconstructed, rebuild and clearing up the power alarm : 

https://files.catbox.moe/nh1wg3.mp4

Tested it just in case, it doesn't touch the air and fire alarms.

</details>

## Changelog
:cl:
fix: Fixed APCs not showing the BSOD after being hacked by a malf AI
fix: Fixed replacing an APC not fixing the power alarm in the area
/:cl:

### CODE REVIEW

File: code/modules/power/apc.dm [hunk 1]
```diff
@@ -196,12 +197,27 @@
...
+	for (var/I in L)
+		if(I == A.name)
+			var/list/alarm = L[I]
+			var/list/srcs  = alarm[3]
+			for(var/origin in srcs)
+				if(origin != source)//We don't want to clear our own alarm, do we
+					area.poweralert(TRUE, origin)
+					srcs -= origin
+			if (srcs.len == 0)
+				L -= I
```
Comment (Penwin0) [inline code/modules/power/apc.dm]: Is this entire proc even necessary? shouldn't `poweralert()` take care of this?

File: code/modules/power/apc.dm [hunk 2]
```diff
@@ -380,7 +398,7 @@
 		if(cell)
 			new_update_state |= UPSTATE_CELL_IN
 
-	else if((obj_flags & EMAGGED) || malfai)
+	else if((obj_flags & EMAGGED) || malfai || malfhack)
```
Comment (Penwin0) [inline code/modules/power/apc.dm]: this is redundant I believe

**Comment by Penwin0** [comment]
This PR is a bit messy, I wish the fixes and the addition had been on seperate ones

File: code/modules/power/apc.dm (hunk 1 above)
Comment (CydiaLamiales) [reply code/modules/power/apc.dm]: Nnnope, poweralert needs an APC as an object, and if the previous APC gets completelly destroyed, it even gets removed from the global APC list sooo there's no way to clear up the power alert tied to it.

**Comment by CydiaLamiales** [comment]
Yeah I feel like I went a bit wild. Should I split these? Won't take long.

**Comment by Penwin0** [comment]
> Yeah I feel like I went a bit wild. Should I split these? Won't take long.

That'd be pretty cool

**Comment by CydiaLamiales** [comment]
Alright, this is now just a bugfix, no more features :tm:

File: code/modules/power/apc.dm [hunk 3]
```diff
@@ -196,12 +196,27 @@
...
 		name = "\improper [get_area_name(area, TRUE)] APC"
 		set_machine_stat(machine_stat | MAINT)
 		update_appearance()
 		addtimer(CALLBACK(src, .proc/update), 5)
+		area.poweralert(FALSE, src)
+
+/obj/machinery/power/apc/proc/clear_previous_power_alarm(obj/source, area/A)
+	var/list/L = GLOB.alarms["Power"]
+	for (var/I in L)
+		if(I == A.name)
```
Comment (itsmeow) [inline code/modules/power/apc.dm]:
```
if(I != A.name)
    continue
```
is cleaner

File: code/modules/power/apc.dm [hunk 4]
```diff
@@ -196,12 +196,27 @@
...
 		operating = FALSE
 		name = "\improper [get_area_name(area, TRUE)] APC"
 		set_machine_stat(machine_stat | MAINT)
 		update_appearance()
 		addtimer(CALLBACK(src, .proc/update), 5)
+		area.poweralert(FALSE, src)
+
+/obj/machinery/power/apc/proc/clear_previous_power_alarm(obj/source, area/A)
+	var/list/L = GLOB.alarms["Power"]
+	for (var/I in L)
```
Comment (itsmeow) [inline code/modules/power/apc.dm]: please use better variable names 😔

File: code/modules/power/apc.dm [hunk 5]
```diff
@@ -196,12 +196,27 @@
...
+		if(I == A.name)
+			var/list/alarm = L[I]
+			var/list/srcs  = alarm[3]
+			for(var/origin in srcs)
```
Comment (itsmeow) [inline code/modules/power/apc.dm]: change to `sources`. `srcs` implies `src` which is totally different

File: code/modules/power/apc.dm [hunk 6]
```diff
@@ -196,12 +196,27 @@
...
+				if(origin != source)//We don't want to clear our own alarm, do we
+					area.poweralert(TRUE, origin)
+					srcs -= origin
+			if (srcs.len == 0)
```
Comment (itsmeow) [inline code/modules/power/apc.dm]: nitpicky but use `!length(srcs)`

**Review by itsmeow — CHANGES_REQUESTED** [review]

**Review by itsmeow — APPROVED** [review]

**Review by PowerfulBacon — APPROVED** [review]

**Pull request merged**